### DISCUSSION TITLE
Is it possible to overwrite the "run-mode" per action?

### DISCUSSION THREAD

**Discussion by alfredbez** [captainhook-git/captainhook#194] [Q&A] [answered]
Hi there, I'm currently working on integrating captainhook into our project. We're using a docker-based setup and have a bunch of actions defined already. Some of the actions might be executed on the host machine, but others need to be executed inside docker (e.g. running migrations).

Imagine having a setup like this:

`captainhook.json`
```json
{
  "config": {
    "bootstrap": "vendor/autoload.php",
    "php-path": "php",
    "run-mode": "docker",
    "run-exec": "docker exec cli"
  },
  "pre-commit": {
    "enabled": true,
    "actions": [
      {
        "action": "\\CaptainHook\\App\\Hook\\PHP\\Action\\Linting",
        "options": [],
        "conditions": []
      },
      {
        "action": "vendor/bin/phpcs {$STAGED_FILES|of-type:php}",
        "options": [],
        "conditions": [
          {
            "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileStaged\\OfType",
            "args": ["php"]
          }
        ]
      }
    ]
  },
  "post-merge": {
    "enabled": true,
    "actions": [
      {
        "action": "vendor/bin/console migrate",
        "options": {},
        "conditions": [
          {
            "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileChanged\\Any",
            "args": [
              ["src/Migration/*.php"]
            ]
          }
        ]
      }
    ]
  }
}
```

We could speed things up by running all actions (except the migrate command) on the host machine by introducing this `captainhook.config.json`:
```json
{
  "run-mode": "local",
  "run-exec": ""
}
```

But this doesn't work with the migrate command. Would be nice to have an option like this:
```diff
@@ -30,7 +30,10 @@
     "actions": [
       {
         "action": "vendor/bin/console migrate",
-        "options": {},
+        "options": {
+          "run-mode": "docker",
+          "run-exec": "docker exec cli"
+        },
         "conditions": [
           {
             "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileChanged\\Any",
```

----

Is it already possible somehow to achieve that? Would you be open for an PR that implements that?

**Comment by sebastianfeldmann** [comment, marked as answer]
I would suggest to change the migrate command to something like

  docker exec "php vendor/bin/console migrate…"

That‘s what the Cap’n would have to do anyway, start a ne "sub" process.

If you want to keep both options, so people could still run everything in docker you would need two `Actions` one with the `vendor/bin…` command and one with the `docker exec … ` command and then apply conditions to each. One where you only run in `run-mode` docker and the other that checks for `run-mode` local.

**Reply by alfredbez** [reply]
Thanks, I guess I will then need to define the actions twice, but our `captainhook.json` is already almost 400 lines long 🙈 

Maybe I'll have a look if I can extend that inside our project

**Reply by alfredbez** [reply]
Would it be possible to have a custom config value like this one?

```json
{
  "action": "{$CONFIG|value-of:command-prefix}myCommand",
  "options": [],
  "conditions": []
}
```

We could then have something like this in our default config:
```json
{
  "config": {
    "bootstrap": "vendor/autoload.php",
    "command-prefix": "docker exec cli "
  },
```

And if someone has php properly configured on his host he could overwrite the config in his `captainhook.config.json` like so:
```json
{
  "command-prefix": ""
}
```

We could then add the prefix to all commands that could be executed on the host and just leave the hardcoded prefix in the commands that need to be executed inside docker like so:
```json
{
  "config": {
    "bootstrap": "vendor/autoload.php",
    "php-path": "php",
    "command-prefix": "docker exec cli"
  },
  "pre-commit": {
    "enabled": true,
    "actions": [
      {
        "action": "\\CaptainHook\\App\\Hook\\PHP\\Action\\Linting",
        "options": [],
        "conditions": []
      },
      {
        "action": "{$CONFIG|value-of:command-prefix}vendor/bin/phpcs {$STAGED_FILES|of-type:php}",
        "options": [],
        "conditions": [
          {
            "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileStaged\\OfType",
            "args": ["php"]
          }
        ]
      }
    ]
  },
  "post-merge": {
    "enabled": true,
    "actions": [
      {
        "action": "docker exec cli vendor/bin/console migrate",
        "options": {},
        "conditions": [
          {
            "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileChanged\\Any",
            "args": [
              ["src/Migration/*.php"]
            ]
          }
        ]
      }
    ]
  }
}
```

**Reply by sebastianfeldmann** [reply]
I could add something like 
```
"config": {
  "custom": {
    "whateveryouwant": "some value"
  }
}
```
and enhance the replacement in a way that this works
```
{$CONFIG|value-of:custom--whateveryouwant} // will replace to "some value"
```
This way I'm not cluttering the config settings and give users the freedom to use basically anything in there config files.
What do you think?

**Reply by alfredbez** [reply]
This world be perfect 👍

**Reply by sebastianfeldmann** [reply]
I just released v5.13.0 with the custom settings feature. The only thing I changed from the description above is the placeholder

    {$CONFIG|value-of:custom>>whateveryouwant} // will replace to "some value"

So I used `>>` over `--`
